### PR TITLE
Core factoring

### PR DESCRIPTION
**Pull request by robrix**
This PR:

- [x] Renames all the `Data.*` modules in `semantic-core` to `Core.*`.
- [x] Brings in a dependency on `fused-syntax`, which provides functionality for scope-safe terms using quantified constraints.
- [x] Replaces a huge swath of syntax-related machinery in `semantic-core` with the definitions in `fused-syntax`.
- [x] Refactors the `Readline` effect and carrier considerably to bring it into line with my more recent implementation from `path`.

### CODE REVIEW

File: semantic-core/semantic-core.cabal [hunk 1]
```diff
@@ -67,29 +33,73 @@ library
...
+    Analysis.Typecheck
+    Control.Carrier.Fail.WithLoc
+    Control.Carrier.Readline.Haskeline
+    Control.Effect.Readline
+    Core.Core
+    Core.Core.Parser
+    Core.Core.Pretty
+    Core.File
+    Core.Loc
+    Core.Name
```
Comment (robrix) [inline semantic-core/semantic-core.cabal]: Note the new names, and the absence of `.Scope`, `.Stack`, and `.Term`.

File: semantic-core/semantic-core.cabal [hunk 2]
```diff
@@ -67,29 +33,73 @@ library
...
+    , filepath                    ^>= 1.4
+    , fused-effects               ^>= 0.5
+    , fused-syntax
+    , haskeline                   ^>= 0.7.5
+    , parsers                     ^>= 0.12.10
+    , prettyprinter               ^>= 1.2.1
+    , prettyprinter-ansi-terminal ^>= 1.1.1
+    , semantic-source             ^>= 0
+    , semigroupoids               ^>= 5.3
+    , terminal-size               ^>= 0.3
```
Comment (robrix) [inline semantic-core/semantic-core.cabal]: This is new; it gives us pretty-printing fit to the terminal width.

File: semantic-core/src/Analysis/Eval.hs [hunk 3]
```diff
@@ -39,7 +39,7 @@ eval :: ( Carrier sig m
      -> (Term (Ann Span :+: Core) Name -> m value)
 eval Analysis{..} eval = \case
   Var n -> lookupEnv' n >>= deref' n
-  Term (R c) -> case c of
+  Alg (R c) -> case c of
```
Comment (robrix) [inline semantic-core/src/Analysis/Eval.hs]: `fused-syntax` gives the recursive constructor of `Term` the name `Alg` instead.

File: semantic-core/src/Control/Carrier/Readline/Haskeline.hs [hunk 4]
```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE FlexibleInstances, GeneralizedNewtypeDeriving, MultiParamTypeClasses, ScopedTypeVariables, TypeApplications, TypeOperators, UndecidableInstances #-}
+module Control.Carrier.Readline.Haskeline
```
Comment (robrix) [inline semantic-core/src/Control/Carrier/Readline/Haskeline.hs]: This has been split out of `Control.Effect.Readline`, and significantly streamlined.

File: semantic-core/src/Control/Carrier/Readline/Haskeline.hs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+instance MonadException m => Carrier Readline (ReadlineC m) where
+  eff (Prompt prompt k) = ReadlineC $ do
+    str <- sendM (getInputLine @m (cyan <> prompt <> plain))
+    Line line <- ask
+    local increment (runReadlineC (k line str))
+    where cyan = "\ESC[1;36m\STX"
+          plain = "\ESC[0m\STX"
+  eff (Print doc k) = do
+    s <- maybe 80 Size.width <$> liftIO size
+    liftIO (renderIO stdout (layoutSmart defaultLayoutOptions { layoutPageWidth = AvailablePerLine s 0.8 } (doc <> line)))
```
Comment (robrix) [inline semantic-core/src/Control/Carrier/Readline/Haskeline.hs]: We now print output wrapped to the terminal size.

File: semantic-core/src/Control/Effect/Readline.hs [hunk 6]
```diff
@@ -1,108 +1,30 @@
...
-
-  runReadline prefs settings block
-
-runControlIO :: (forall x . m x -> IO x) -> ControlIOC m a -> m a
-runControlIO handler = runReader (Handler handler) . runControlIOC
-
--- | This exists to work around the 'MonadException' constraint that haskeline entails.
-newtype ControlIOC m a = ControlIOC { runControlIOC :: ReaderC (Handler m) m a }
-  deriving newtype (Applicative, Functor, Monad, MonadIO)
+  = Prompt String (Int -> Maybe String -> m k)
```
Comment (robrix) [inline semantic-core/src/Control/Effect/Readline.hs]: `Line` is now an implementation detail of the carrier; plus we now return the line number along with the text (if any) from the `Prompt` constructor, and 🔥 `AskLine`.

File: semantic-core/src/Control/Effect/Readline.hs [hunk 7]
```diff
@@ -1,108 +1,30 @@
...
+  | Print (Doc AnsiStyle) (m k)
```
Comment (robrix) [inline semantic-core/src/Control/Effect/Readline.hs]: Specializing this to `AnsiStyle` obviates the need for `AnyDoc`.

File: semantic-core/src/Control/Effect/Readline.hs [hunk 8]
```diff
@@ -1,108 +1,30 @@
...
-  eff op = ControlIOC (eff (R (handleCoercible op)))
+prompt :: (Member Readline sig, Carrier sig m) => String -> m (Int, Maybe String)
+prompt p = send (Prompt p (curry pure))
 
-instance (Carrier sig m, MonadIO m) => MonadException (ControlIOC m) where
-  controlIO f = ControlIOC $ do
-    handler <- ask
-    liftIO (f (RunIO (fmap pure . runHandler handler)) >>= runHandler handler)
+print :: (Carrier sig m, Member Readline sig) => Doc AnsiStyle -> m ()
+print s = send (Print s (pure ()))
```
Comment (robrix) [inline semantic-core/src/Control/Effect/Readline.hs]: `print` always ends with a newline, so we don’t need `println`.

File: semantic-core/src/Core/Core/Pretty.hs [hunk 9]
```diff
@@ -99,21 +100,3 @@ prettyCore style = precBody . go . fmap name
...
-prec :: Int -> a -> Prec a
-prec = Prec . Just
-
-atom :: a -> Prec a
-atom = Prec Nothing
-
-withPrec :: Int -> Prec AnsiDoc -> AnsiDoc
-withPrec d (Prec d' a)
-  | maybe False (d >) d' = parens a
-  | otherwise            = a
```
Comment (robrix) [inline semantic-core/src/Core/Core/Pretty.hs]: This is defined by `fused-syntax`, so we don’t need to define it ourselves any more.

File: semantic-core/src/Analysis/Analysis.hs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  , abstract  :: (term -> m value) -> Name -> term -> m value
+  , apply     :: (term -> m value) -> value -> value -> m value
+  , unit      :: m value
+  , bool      :: Bool -> m value
+  , asBool    :: value -> m Bool
+  , string    :: Text -> m value
+  , asString  :: value -> m Text
+  , record    :: [(Name, value)] -> m value
+  , (...)     :: address -> Name -> m (Maybe address)
+  }
```
Comment (robrix) [inline semantic-core/src/Analysis/Analysis.hs]: I’ve moved `Analysis` here in support of separating out the `Analysis` & `Core` hierarchies completely.

File: semantic-core/src/Analysis/Analysis.hs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+module Analysis.Analysis
+( Analysis(..)
+) where
+
+import Core.Name
+import Data.Text (Text)
+
+-- | A record of functions necessary to perform analysis.
+--
+-- This is intended to be replaced with a selection of algebraic effects providing these interfaces and carriers providing reusable implementations.
```
Comment (robrix) [inline semantic-core/src/Analysis/Analysis.hs]: As mentioned here, `Analysis` has only ever been defined for expedience’s sake; I intend to split it into effects and carriers in a future PR.

File: .travis.yml [hunk 12]
```diff
@@ -36,7 +36,7 @@ install:
 script:
 - cabal new-build
 - cabal new-run semantic:test
-- cabal new-run semantic-core:spec
```
Comment (patrickt) [inline .travis.yml]: Yes. Thank you.

File: cabal.project [hunk 13]
```diff
@@ -33,3 +33,8 @@ source-repository-package
   type: git
   location: https://github.com/tclem/proto-lens-jsonpb
   tag: e4d10b77f57ee25beb759a33e63e2061420d3dc2
+
+source-repository-package
+  type: git
+  location: https://github.com/antitypical/fused-syntax.git
+  tag: 5b7512db962d5b3f973002615b8bc86ab074d5aa
```
Comment (patrickt) [inline cabal.project]: We’ll have to do a Hackage release of this eventually, but this is fine for now.

Comment (robrix) [reply cabal.project]: Yep. I’m mostly held up debating whether to keep or 🔥 the `Analysis` class it has.

I’m also definitely going to rename it.

**Pull request merged**